### ISSUE TITLE
International characters not rendered correctly

### ISSUE THREAD

**Issue by nasko** [bookdown/Bookdown.Bookdown#32]
I'm having issues with Cyrillic characters.

This is my content file `overview.md`:

``` markdown
# Zend Framework 2

![ZF 2 Logo](images/zf2-elephant.png)

## План

* Въведение
* MVC
* Конфигурация
* ServiceManager
* PluginManager
* EventManager
* Модули
* Маршрутизация
```

I can verify that the file is saved with the proper encoding like this:

``` bash
$ file _bookdown/zend-framework-2/overview.md 
_bookdown/zend-framework-2/overview.md: UTF-8 Unicode text
```

When I render the html, I can see that Cyrillic characters used in headings are rendered as non-printable gibberish, while Cyrillic characters used in the page content, are rendered as html entities:

Contents of rendered `index.html`:

``` html
<!-- snip -->

<h1>Zend Framework 2</h1>
<dl>
<dt>1. <a href="./overview.html#1">Zend Framework 2</a></dt>
<dd><dl>
<dt>1.1. <a href="./overview.html#1.1">ÐÐ»Ð°Ð½</a></dt>
</dl></dd>
<dt>2. <a href="./intro.html#2">ÐÑÐ²ÐµÐ´ÐµÐ½Ð¸Ðµ</a></dt>
<dd><dl>
<dt>2.1. <a href="./intro.html#2.1">ÐÐ¾Ð²Ð° Ð°ÑÑÐ¸ÑÐµÐºÑÑÑÐ°</a></dt>
<dt>2.2. <a href="./intro.html#2.2">Ð¡ÑÐ²Ð¼ÐµÑÑÐ¸Ð¼ Ñ PSR-2</a></dt>
</dl></dd>
</dl>

<!-- snip -->

```

This is how the actual content page is rendered in `overview.html`:

``` html

<!-- snip -->
<h1 id="1">1. Zend Framework 2</h1>
<p><img src="images/zf2-elephant.png" alt="ZF 2 Logo"></p>
<h2 id="1.1">1.1. &ETH;&#159;&ETH;&raquo;&ETH;&deg;&ETH;&frac12;</h2>
<ul>
<li>&ETH;&#146;&Ntilde;&#138;&ETH;&sup2;&ETH;&micro;&ETH;&acute;&ETH;&micro;&ETH;&frac12;&ETH;&cedil;&ETH;&micro;_</li>
<li>MVC</li>
<li>&ETH;&#154;&ETH;&frac34;&ETH;&frac12;&Ntilde;&#132;&ETH;&cedil;&ETH;&sup3;&Ntilde;&#131;&Ntilde;&#128;&ETH;&deg;&Ntilde;&#134;&ETH;&cedil;&Ntilde;&#143;_</li>
<li>ServiceManager</li>
<li>PluginManager</li>
<li>EventManager</li>
<li>&ETH;&#156;&ETH;&frac34;&ETH;&acute;&Ntilde;&#131;&ETH;&raquo;&ETH;&cedil;_</li>
<li>&ETH;&#156;&ETH;&deg;&Ntilde;&#128;&Ntilde;&#136;&Ntilde;&#128;&Ntilde;&#131;&Ntilde;&#130;&ETH;&cedil;&ETH;&middot;&ETH;&deg;&Ntilde;&#134;&ETH;&cedil;&Ntilde;&#143;_</li>
</ul>

<!-- snip -->

```

I'm attaching the sources folder: `_bookdown` and the rendered target folder: `zend-framework-2` compressed as ZIP archives.

Might be, I'm not using the tool properly, in which case I'd be quite thankful for any advise you might offer!

[_bookdown.zip](https://github.com/bookdown/Bookdown.Bookdown/files/106315/_bookdown.zip)
[zend-framework-2.zip](https://github.com/bookdown/Bookdown.Bookdown/files/106314/zend-framework-2.zip)

**Comment by sandrokeil** [comment]
Maybe it's a PHP configuration issue? If I generate the HTML I have something like this, which is correct.

``` html
<h2 id="1.1">1.1. &#1055;&#1083;&#1072;&#1085;</h2>
<ul>
<li>&#1042;&#1098;&#1074;&#1077;&#1076;&#1077;&#1085;&#1080;&#1077;</li>
<li>MVC</li>
<li>&#1050;&#1086;&#1085;&#1092;&#1080;&#1075;&#1091;&#1088;&#1072;&#1094;&#1080;&#1103;</li>
<li>ServiceManager</li>
<li>PluginManager</li>
<li>EventManager</li>
<li>&#1052;&#1086;&#1076;&#1091;&#1083;&#1080;</li>
<li>&#1052;&#1072;&#1088;&#1096;&#1088;&#1091;&#1090;&#1080;&#1079;&#1072;&#1094;&#1080;&#1103;</li>
</ul>
```

I've used my Docker Bookdown image with PHP 5.6 but have no special settings.

**Comment by nasko** [comment]
With the luxury of having unicode support, why would you consider rendering of Cyrillic characters as html entities correct? 
Besides, if you type in Cyrillic chars in the book title, you'll note that they're rendered just fine in the target html.

**Comment by sandrokeil** [comment]
Yes that's right. I think it's a problem of the conversion process from Markdown to HTML. This is done by [league/commonmark](https://github.com/thephpleague/commonmark). You can try to use this library directly to ensure it's not a problem of Bookdown. You can also debug the Markdown conversion process of this library or open an issue on this project to get more insights or help.

**Comment by pmjones** [comment]
@colinodell Any thoughts on this issue?

**Comment by nasko** [comment]
Today I took a cursory look at league/commonmark's issues and noted an entry related to Russian characters but when I browsed their tests I was left with the impression that they had it covered. 

Tomorrow I'll dig in more thoroughly. 

Additionally, I'll investigate the few occasions where I spotted `substr` and `strlen` invocations in the bookdown source code

**Comment by colinodell** [comment]
Running it through `league/commonmark` (via the CLI binary with default configuration) produces the correct result:

![image](https://cloud.githubusercontent.com/assets/202034/12631737/09728608-c521-11e5-89dd-3afa8e03f1c8.png)

We do fully support UTF-8, and it seems to work fine.

I agree with @nasko, this issue might be caused by using string functions which don't play nicely with multi-byte characters. `league/commonmark` has a hard dependency on the `mbstring` extension, so you should be okay to use those `mb_` functions instead.

**Comment by sandrokeil** [comment]
@pmjones I think it's a problem of DomNode in the HeadingsProcess. Links are rendered correct.

**Comment by pmjones** [comment]
@sandrokeil Ah, that sounds likely. I wonder if this is a fix: http://php.net/manual/en/domdocument.loadhtml.php#95251

If so, it would go in HeadingsProcess::loadDomDocument().

Thoughts?

**Comment by pmjones** [comment]
There's also http://stackoverflow.com/questions/8218230/php-domdocument-loadhtml-not-encoding-utf-8-correctly (see the second answer about saveHtml() not saving correctly).

**Comment by colinodell** [comment]
@pmjones That first link (https://secure.php.net/manual/en/domdocument.loadhtml.php#95251) is exactly what we're doing in league/html-to-markdown to ensure UTF-8 chars don't get all garbled:

``` php
$document = new \DOMDocument();
$document->loadHTML('<?xml encoding="UTF-8">' . $html);
$document->encoding = 'UTF-8';
```

**Comment by harikt** [comment]
I thought this was fixed with #13 . The issue was also reported in #12 .

**Comment by nasko** [comment]
I'm somewhat confused and am not sure how exactly to proceed. The source code concerned in this issue is different in the master branch as compared to the develop branch. 

Because I intended to use `bookdown/bookdown` in a production application, I initially tried to use the master branch and that's where I encountered this issue. Thanks to the comments of @pmjones, @colinodell and @sandrokeil, I managed to patch the source code, such that it worked just fine. Sorry for pasting diff output here, but this is the only way I can show you what I had to change:

_Note: This is valid for the current master branch_

``` Diff
$ git diff src/Process/Headings/HeadingsProcess.php
diff --git a/src/Process/Headings/HeadingsProcess.php b/src/Process/Headings/HeadingsProcess.php
index d303fba..7b683bf 100644
--- a/src/Process/Headings/HeadingsProcess.php
+++ b/src/Process/Headings/HeadingsProcess.php
@@ -93,7 +93,7 @@ class HeadingsProcess implements ProcessInterface
     {
         $this->doc = new DomDocument();
         $this->doc->formatOutput = true;
-        $this->doc->loadHtml($this->html, LIBXML_HTML_NODEFDTD);
+        $this->doc->loadHtml('<?xml encoding="UTF-8">' . $this->html, LIBXML_HTML_NODEFDTD);
     }

     protected function processHeadingNodes()
@@ -187,12 +187,12 @@ class HeadingsProcess implements ProcessInterface
     protected function setHtmlFromDomDocument()
     {
         // retain the modified html
-        $this->html = trim($this->doc->saveHtml());
+        $this->html = trim($this->doc->saveHtml($this->doc->documentElement));

         // strip the html and body tags added by DomDocument
         $this->html = substr(
             $this->html,
-            strlen('<html><body>'),
+            strlen('<?xml encoding="UTF-8"><html><body>'),
             -1 * strlen('</body></html>')
         );
```

However, when I tried to create a pull request, I read the guidelines and I knew that I had to create the pull request based on the develop branch. When I tried to apply the same changes to the develop branch, I noticed that the code had changed - with respect to the `$this->doc->loadHtml()` invocation in `HeadingsProcess::loadDomDocument()`. 

Apparently, the changes I did to that method were not needed in the develop branch as that was already taken care about, via the  `mb_convert_encoding()` call. 

I've forked the repository and created a fix branch off of `develop`, where I committed a different patch: nasko/Bookdown.Bookdown@e9c25484b2eed4bce608ad89a02c3b5c9f185f14

I'm not eager to create a pull request, because I'm not quite skilled with respect to tests, and that was one of the things mentioned in the contribution guidelines.

Any thoughts are welcome!

**Comment by nasko** [comment]
In my opinion, what happened was that the cumulative effect of two distinct issues hit me:
- International characters were being rendered as gibberish - apparently this has been fixed in the `develop` branch by #13 (thanks, @harikt for the heads up!), but this has not yet made it the `master` branch.
  
  In my patch listed in my previous comment, I've tackled this, based on the code @colinodell showed:

``` Diff
-        $this->doc->loadHtml($this->html, LIBXML_HTML_NODEFDTD);
+        $this->doc->loadHtml('<?xml encoding="UTF-8">' . $this->html, LIBXML_HTML_NODEFDTD);
```

  Note, that I had to account for this extra chunk of markup later in `setHtmlFromDomDocument()`:

``` php
        $this->html = substr(
            $this->html,
            strlen('<?xml encoding="UTF-8"><html><body>'),
            -1 * strlen('</body></html>')
        );
```

  Anyway, I think the `mb_convert_encoding()` approach is better and less hackish.
- Some of the content from the resource files was being rendered as html entities in the resulting target file.
  
  To fix this, I used the [comment](http://stackoverflow.com/a/20675396) that @pmjones referred to, in my nasko/Bookdown.Bookdown@e9c25484b2eed4bce608ad89a02c3b5c9f185f14 commit
